### ISSUE TITLE
Scale for text/features

### ISSUE THREAD

**Issue by camilleanne** [mapbox/mapbox-studio-classic#458]
Capturing a conversation with @amyleew & @springmeyer. 

Create a separate scale value for text, halos, and other features that increase in size during tile size scaling in a way that is not desirable for print.

![scale](https://cloud.githubusercontent.com/assets/3952537/3347052/d6e3178c-f8d2-11e3-88e7-8f599a8f1f98.png)

**Comment by yhahn** [comment]
Hmm, let's talk about this next week. I'm strongly opposed to an additional value here because it undermines what the actual carto CSS styles mean (basically, nothing -- someone can come mess with the "text/line" scale and completely alter your design).

To address these use cases I think either the fractional scale factor or making adjustments directly to styles makes more sense.

**Comment by amyleew** [comment]
@yhahn text/halo/lines are consistently printing too large in my testing... because web text looks best at around 14pt but printed text looks best around 8-9pt. i asked @camilleanne about the scale factor tool from TM1, thinking a workflow enhancement to the TM2 print tool could be a function/slider that will globally scale down text/halo/lines without having to re-code each instance in CSS.

**Comment by yhahn** [comment]
@camilleanne @amyleew I didn't do a very good job of explaining the background behind this response -- here's a bit more context.

### In vector tile-based design, tile-size is the critical unit

Unlike when designing in a graphics editor or on the web a map designer is actually working with the constraint of map tiles.

When a CartoCSS style specifies that certain text labels are `16` at a certain zoom level, it is in a sense defining something much more important than text size in pixels -- it's defining a relationship between the text size, tile size, and density of data within that tile. In the example below, the text sizes are chosen in a way where there's a certain amount of spacing between labels and a certain proportion of text to the entire tile size.

![scale1](https://cloud.githubusercontent.com/assets/83384/3350573/ec1986d0-f9be-11e3-93fd-874828d519aa.png)

_While at first you think you're making a decision about text size in pixels, it's the relationship to other sizes (tile, other labels, data density, etc.) that is as important if not more important._

### When tile size follow scale size, relationships are preserved

![scale2](https://cloud.githubusercontent.com/assets/83384/3350574/f2dded30-f9be-11e3-9bce-8b3465c2161b.png)

In @camilleanne's work here one of the main things we were accomplishing was to make scale factor mean a much more consistent thing to map designers. No matter what the scale factor of a map is, the relationship established by styles would be preserved. This meant very straightforward rules and a strict distinction between which tool was used to accomplish what:
- Want to alter the design (sizes of text relative to line widths, tile size, etc.) of your map? Alter the carto css.
- Want to alter the number of pixels/pixel density used to render your map? Alter the scale factor.

### Two scale factors...

... would basically undo this previous work (https://github.com/mapbox/tilelive-vector/pull/57) making it possible again to alter scale factor independently of tile size. Here's an example of the kinds of problems this used to cause that we'd be allowing again:

![scale3](https://cloud.githubusercontent.com/assets/83384/3350575/f649dbc8-f9be-11e3-8f71-b42df7af7f63.png)
- We've scaled up label size, but because none of the tile data density was scaled,
- we lost the "Apples" label which now collides with 14th St and,
- the spacing between labels is much more cramped.

**In other words we altered the design of the map in ways we did not expect to. From a UX standpoint, advertising scale as a way to make design decisions about a map is deceptive.**

It has the appearance of a magic setting that's going to give the user smaller or bigger labels but in reality if it's not coordinated with tile size and data density a lot of other bad things are going to come along with it.

---

I think for now print maps that need smaller labels should just use smaller labels in cartocss. We can use variables and other tricks to make this less painful, but it's the right tool for the job -- you're making a design decision, not an output/format one.

**Comment by yhahn** [comment]
Last thought here is -- imagine a huge map poster printed that looks like this:

![postermap](https://cloud.githubusercontent.com/assets/83384/3350591/6f280ece-f9c0-11e3-816c-7eccf76c2775.png)

Is this a
- a 200,000x500,000px scale poster with cartocss that defines 30,000px text size? or is it
- a 200,000x500,000px scale poster with cartocss that defines 30px text size but at scale=1000?
- does it matter?

I'm wondering if we should think about scale factor as not being tied to dpi/ppi at all in the print context (it's still very dpi/ppi bound in the retina screen context). We could see it as both a way to print higher density images (if we want) but also a way to just scale up/down all aspects of a map (text, lines, etc).

**Comment by ghost** [comment]
> - Want to alter the design (sizes of text relative to line widths, tile size, etc.) of your map? Alter the carto css.
> - Want to alter the number of pixels/pixel density used to render your map? Alter the scale factor.
> 
> ...
> I think for now print maps that need smaller labels should just use smaller labels in cartocss.

@amyleew and @camilleanne I'm not sure if I ever articulated it this plainly ^^ when we were discussing things like this last week, but I'm inclined to agree that things ought to work this way (not sure to what extent they do now, but that's a different issue).

The one point that I'd add is, I think it would be great to have the UI provide assistance for picking the correct scale factor based on the intended bounding box and output format. I'm thinking presets like scale=_x_ for normal web tiles, _2x_ for retina tiles, _y_ for letter size, _z_ for poster size, plus a "custom..." option to type in an exact value.

**Comment by amyleew** [comment]
@yhahn thanks for the clarification! i'll look to carto css as the best method for text/line/label control. and share use case / workflow examples as they come up.

**Comment by camilleanne** [comment]
@yhahn thanks for a more detailed explanation. Even after working on this I realize I have a very limited understanding of the larger paradigms at work and choices already made.

I agree that these changes should be in the cartocss. It seems easier to add a slider, but that's deceptive. It would be nice to see some patterns in carto come out of this work for print and web, maybe relying heavily on the variables in cartocss to toggle between print appropriate feature sizing and web appropriate in the same style.

**Comment by amyleew** [comment]
@camilleanne @yhahn completely agree. perhaps, a print.mss and web.mss which include the elements that are change depending on respective final output. i'll continue to work towards those use case scenarios as i print.

**Comment by yhahn** [comment]
:+1:
